File: internal/connhealth/connhealth_test.go
package connhealth

import (
	"testing"
	"time"
)

func TestTracker_InitialSnapshot(t *testing.T) {
	tr := New()
	s := tr.Snapshot(time.Unix(0, 0))
	if s.State != StateClosed {
		t.Fatalf("initial state = %q, want %q", s.State, StateClosed)
	}
	if s.RTT.LastMS != nil {
		t.Fatalf("initial RTT non-nil")
	}
	if s.Reconnect.CountLastHour != 0 {
		t.Fatalf("initial reconnect count = %d", s.Reconnect.CountLastHour)
	}
}

func TestTracker_RecordsRTTAndComputesPercentiles(t *testing.T) {
	tr := New()
	tr.SetState(StateConnected, time.Unix(0, 0))
	for i, ms := range []int{50, 60, 70, 80, 90, 100, 110, 120, 130, 140} {
		tr.OnPongRTT(ms, time.Unix(int64(i), 0))
	}
	s := tr.Snapshot(time.Unix(10, 0))
	if got := *s.RTT.LastMS; got != 140 {
		t.Fatalf("last RTT = %d, want 140", got)
	}
	if got := *s.RTT.P50MS; got < 90 || got > 100 {
		t.Fatalf("p50 = %d, want 90..100", got)
	}
	if got := *s.RTT.P95MS; got < 130 || got > 140 {
		t.Fatalf("p95 = %d, want 130..140", got)
	}
	if len(s.RTTSamples) != 10 {
		t.Fatalf("samples len = %d, want 10", len(s.RTTSamples))
	}
}

func TestTracker_RTTRingEvictsOldest(t *testing.T) {
	tr := New()
	tr.SetState(StateConnected, time.Unix(0, 0))
	for i := 0; i < RTTRingSize+5; i++ {
		tr.OnPongRTT(i, time.Unix(int64(i), 0))
	}
	s := tr.Snapshot(time.Unix(0, 0))
	if len(s.RTTSamples) != RTTRingSize {
		t.Fatalf("ring size = %d, want %d", len(s.RTTSamples), RTTRingSize)
	}
	// Oldest sample should be i=5 (we wrote 65 items into a 60-slot ring).
	first := s.RTTSamples[0]
	if first.RTTMS != 5 {
		t.Fatalf("ring[0].rtt = %d, want 5", first.RTTMS)
	}
}

func TestTracker_StateTransitionsLogReconnects(t *testing.T) {
	tr := New()
	tr.SetState(StateConnecting, time.Unix(0, 0))
	tr.SetState(StateConnected, time.Unix(1, 0))
	tr.SetState(StateReconnecting, time.Unix(60, 0))
	tr.SetReconnectReason("ws_close_1006", time.Unix(60, 0))
	tr.SetState(StateConnected, time.Unix(62, 0))
	s := tr.Snapshot(time.Unix(120, 0))
	if s.Reconnect.CountLastHour != 1 {
		t.Fatalf("reconnect count = %d, want 1", s.Reconnect.CountLastHour)
	}
	if len(s.Reconnect.History) != 1 {
		t.Fatalf("history len = %d", len(s.Reconnect.History))
	}
	h := s.Reconnect.History[0]
	if h.Reason != "ws_close_1006" {
		t.Fatalf("reason = %q", h.Reason)
	}
	if h.DurationMS != 2000 {
		t.Fatalf("downtime = %d ms, want 2000", h.DurationMS)
	}
}

func TestTracker_ReconnectCountWindowsAtOneHour(t *testing.T) {
	tr := New()
	tr.SetState(StateConnected, time.Unix(0, 0))
	// Two reconnects ~ at t=100/200s (outside the 1h window from t=5500).
	tr.SetState(StateReconnecting, time.Unix(100, 0))
	tr.SetState(StateConnected, time.Unix(101, 0))
	tr.SetState(StateReconnecting, time.Unix(200, 0))
	tr.SetState(StateConnected, time.Unix(201, 0))
	// One inside the window.
	tr.SetState(StateReconnecting, time.Unix(5000, 0))
	tr.SetState(StateConnected, time.Unix(5001, 0))
	s := tr.Snapshot(time.Unix(5500, 0))
	if s.Reconnect.CountLastHour != 1 {
		t.Fatalf("reconnect count = %d, want 1 (5000s is inside, 100/200s are outside)", s.Reconnect.CountLastHour)
	}
}

func TestTracker_BytesEMA(t *testing.T) {
	tr := New()
	tr.SetState(StateConnected, time.Unix(0, 0))
	// 1000 bytes/s for 30 seconds (so the EMA has time to converge).
	for i := 0; i < 30; i++ {
		tr.OnBytesIn(1000, time.Unix(int64(i), 0))
		tr.OnBytesOut(500, time.Unix(int64(i), 0))
		tr.Tick(time.Unix(int64(i+1), 0)) // tick at end of bucket
	}
	s := tr.Snapshot(time.Unix(30, 0))
	if s.Bytes.InPerSec < 800 || s.Bytes.InPerSec > 1100 {
		t.Fatalf("in_per_sec = %d, want ~1000", s.Bytes.InPerSec)
	}
	if s.Bytes.OutPerSec < 400 || s.Bytes.OutPerSec > 550 {
		t.Fatalf("out_per_sec = %d, want ~500", s.Bytes.OutPerSec)
	}
}

func TestTracker_SeqGapsCounter(t *testing.T) {
	tr := New()
	tr.OnSeqGap()
	tr.OnSeqGap()
	tr.OnSeqGap()
	if got := tr.Snapshot(time.Unix(0, 0)).SeqGaps; got != 3 {
		t.Fatalf("seq_gaps = %d, want 3", got)
	}
}

File: web/tests/unit/shared/connhealth/connhealth.test.ts
import { describe, it, expect, beforeEach, vi } from 'vitest'
import { Tracker, RTT_RING_SIZE } from '@shared/connhealth/connhealth'

describe('Tracker', () => {
  let tr: Tracker
  beforeEach(() => {
    tr = new Tracker()
  })

  it('starts in the closed state with empty RTT', () => {
    const s = tr.snapshot(0)
    expect(s.state).toBe('closed')
    expect(s.rtt.last_ms).toBeNull()
    expect(s.rtt_samples).toEqual([])
    expect(s.reconnect.count_last_hour).toBe(0)
  })

  it('records RTT samples and computes p50/p95', () => {
    tr.setState('connected', 0)
    for (let i = 0; i < 10; i++) {
      tr.onPongRTT(50 + i * 10, i * 1000)
    }
    const s = tr.snapshot(10_000)
    expect(s.rtt.last_ms).toBe(140)
    expect(s.rtt.p50_ms!).toBeGreaterThanOrEqual(90)
    expect(s.rtt.p50_ms!).toBeLessThanOrEqual(100)
    expect(s.rtt.p95_ms!).toBeGreaterThanOrEqual(130)
    expect(s.rtt.p95_ms!).toBeLessThanOrEqual(140)
    expect(s.rtt_samples.length).toBe(10)
  })

  it('evicts oldest RTT samples past RTT_RING_SIZE', () => {
    tr.setState('connected', 0)
    for (let i = 0; i < RTT_RING_SIZE + 5; i++) {
      tr.onPongRTT(i, i * 1000)
    }
    const s = tr.snapshot(0)
    expect(s.rtt_samples.length).toBe(RTT_RING_SIZE)
    expect(s.rtt_samples[0].rtt_ms).toBe(5)
  })

  it('tracks reconnect events with duration and reason', () => {
    tr.setState('connected', 0)
    tr.setState('reconnecting', 60_000)
    tr.setReconnectReason('ws_close_1006', 60_000)
    tr.setState('connected', 62_000)
    const s = tr.snapshot(120_000)
    expect(s.reconnect.count_last_hour).toBe(1)
    expect(s.reconnect.history).toHaveLength(1)
    expect(s.reconnect.history[0].reason).toBe('ws_close_1006')
    expect(s.reconnect.history[0].duration_ms).toBe(2000)
  })

  it('windows reconnect count at one hour', () => {
    tr.setState('connected', 0)
    tr.setState('reconnecting', 100_000)
    tr.setState('connected', 101_000)
    tr.setState('reconnecting', 200_000)
    tr.setState('connected', 201_000)
    tr.setState('reconnecting', 5_000_000)
    tr.setState('connected', 5_001_000)
    const s = tr.snapshot(5_500_000)
    expect(s.reconnect.count_last_hour).toBe(1)
  })

  it('computes byte EMAs after sustained ticks', () => {
    tr.setState('connected', 0)
    for (let i = 0; i < 30; i++) {
      tr.onBytesIn(1000, i * 1000)
      tr.onBytesOut(500, i * 1000)
      tr.tick((i + 1) * 1000)
    }
    const s = tr.snapshot(30_000)
    expect(s.bytes.in_per_sec).toBeGreaterThanOrEqual(800)
    expect(s.bytes.in_per_sec).toBeLessThanOrEqual(1100)
    expect(s.bytes.out_per_sec).toBeGreaterThanOrEqual(400)
    expect(s.bytes.out_per_sec).toBeLessThanOrEqual(550)
  })

  it('counts seq gaps via onSeqGap', () => {
    tr.onSeqGap()
    tr.onSeqGap()
    tr.onSeqGap()
    expect(tr.snapshot(0).seq_gaps).toBe(3)
  })

  it('emits onChange listeners on state transitions', () => {
    const listener = vi.fn()
    tr.onChange(listener)
    tr.setState('connecting', 0)
    tr.setState('connected', 100)
    expect(listener).toHaveBeenCalledTimes(2)
  })
})
